# West Coast Dream Team ATAC-seq analysis of mCRPC
_This code repository reproduces the analyses described in Shrestha et al., Cancer Research, 2024._

### Citation
Shrestha R. et al. An Atlas of Accessible Chromatin in Advanced Prostate Cancer Reveals the Epigenetic Evolution during Tumor Progression. Cancer Research. 2024

### Abstract
> Metastatic castration-resistant prostate cancer (mCRPC) is a lethal disease that resists therapy targeting androgen signaling, the primary driver of prostate cancer. mCRPC resists androgen receptor (AR) inhibitors by amplifying AR signaling or by evolving into therapy-resistant subtypes that do not depend on AR. Elucidation of the epigenetic underpinnings of these subtypes could provide important insights into the drivers of therapy resistance. In this study, we produced chromatin accessibility maps linked to the binding of lineage-specific transcription factors (TF) by performing ATAC sequencing on 70 mCRPC tissue biopsies integrated with transcriptome and whole genome sequencing. mCRPC had a distinct global chromatin accessibility profile linked to AR function. Analysis of TF occupancy across accessible chromatin revealed 203 TFs associated with mCRPC subtypes. Notably, ZNF263 was identified as a putative prostate cancer TF with a significant impact on gene activity in the double-negative (AR- neuroendocrine-) subtype, potentially activating MYC targets. Overall, this analysis of chromatin accessibility in mCRPC provides valuable insights into epigenetic changes that occur during progression to mCRPC.

### For any further inquries
>Please direct correspondence related to this manuscript to Felix Feng, UCSF.  
>Please direct inquiries related to the data in this manuscript to David Quigley, UCSF.

<br/>

## Download Dataset 
### 1. Raw sequencing data: 
The raw sequencing data generated or used in this study are available under the following EGA dataset accession:

| Data Type | Accession Number | URL                                              |
|-----------|------------------|--------------------------------------------------|
| ATAC-seq  | EGAS00001006698  | https://ega-archive.org/datasets/EGAS00001006698 |
| RNA-seq   | EGAD00001008487  | https://ega-archive.org/datasets/EGAD00001008487 |
| RNA-seq   | EGAD00001008991  | https://ega-archive.org/datasets/EGAD00001008991 |
| RNA-seq   | EGAD00001009065  | https://ega-archive.org/datasets/EGAD00001009065 |


### 2. Processed data used in this study:
The processed data and binary files required to reproduce the analyses and figures in this study are too large to publish on GitHub. These files can be obtained at Zenodo at DOI:

DOI 10.5281/zenodo.11660490

[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.11660491.svg)](https://doi.org/10.5281/zenodo.11660491)